### DISCUSSION TITLE
Performance

### DISCUSSION THREAD

**Discussion by conartist6** [iter-tools/iter-tools#442] [Ideas]
The biggest roadblock to wider adoption of iter-tools and other libraries like it is that it's, well, slow. It's hard to tell people they should be using these methods lots of places unless we can say they're at least reasonably fast. This thread is for brainstorming ideas on how to improve perf.

**Comment by conartist6** [comment]
One idea i've been tossing around is to make a lower level iteration protocol and wrap it. In general I've been building the infra to do something like this, building a clearly differentiated implementation and wrapper layer. One way to boost perf I think would be to swap out the implementation layer for a different iteration protocol.

The current iteration protocol involves `{ item, value }` step objects. There are a lot of these objects, roughly `t * n` of them where `t` is the number of chained transforms and `n` is the number of steps in the source iterable. They're relatively cheap to allocate I think but they all have to be swept by the garbage collector and that's just a lot of work. In general it is consistent with the iter-tools philosophy not to shy away from creating a lot of small objects as they can be cleaned up in minor collections which are less disruptive to responsive apps, but that doesn't mean we can't still degrade perf meaningfully by requiring the GC to be working overtime.

It seems to me that I could swap the internal iteration protocol for one that does not use step objects and instead signals the end of an iteration using a sentinel value, probably a symbol. This would only eliminate the possibility of steps like `{ value: true, done: true}`, but iter-tools doesn't use return values for anything, and is unlikely to in the future. This way we could make only `n` or `2n` step objects no matter how many transforms were present. Unfortunately this would lose us the ability to factor the implementation in terms of generator functions. Right now generator functions are one of the things that are slow, but that may not always be the case. Either way we could still design with generator functions and then transpile them down to something else.

**Reply by hillin** [reply]
> The current iteration protocol involves { item, value } step objects. There are a lot of these objects, roughly t * n of them where t is the number of chained transforms and n is the number of steps in the source iterable.

Maybe use an object pool?

**Reply by conartist6** [reply]
I'd be much happier just not using objects at all, but the point is that generator functions don't give me a chance to use any other behavior but their own. They always create new step objects.

For now what generators get me is worth what they cost me I think. That's because of the generational garbage collection strategy which makes collecting those objects pretty cheap. It's cheap because the smallest, youngest objects are assumed to be short-lived and are allocated in a Gen-1 memory pool, only a few MB in size. Every so often as the Gen-1 pool fills up its contents are marked and swept, a whole new Gen-1 pool is allocated, and surviving objects are copied into it. In this way the engines make it mostly fine to use a lot of small objects. A recent benchmark of the [@iter-tools/regex](https://github.com/iter-tools/regex) engine in which I thought small object allocation and collection might be one of the more major costs showed that it was consuming 200ms in a 4 second run, which to me was not worth attempting to optimize even though in that code I really could use an object pool.

Finally I think it fairly likely that engines themselves will eventually optimize these objects away! If a `for .. of` loop is calling `generator.prototype.next()` generated by a generator then step objects is a completely hidden detail of the implementation. Since user code has no access to that step object at all the engine could completely optimize away the need to create it e.g. by reading from the `generator[[value]]` and `generator[[done]]` internal slots directly.

So to sum up, it's just not worth it. There'd be a lot of cost, there's not enough to gain. and moving away from generators in order to pool objects ourselves might actually make the code slower in the long run by removing the engine's ability to optimize.

**Reply by hillin** [reply]
Of course, it's just a suggestion that if profiling shows the step object allocation/collection is a hot code path, and it's not clear how and when (and whether) would the JS engines optimize it, and there is no better optimization choices, using an object pool would be a good idea. Historically they worked very well and significantly boosted performance in several projects of mine.

**Reply by conartist6** [reply]
You make it sound very simple, but again you're talking about a huge, huge change. All the code in the library would need to be written in a different paradigm! I won't completely rule it out, but I'd need to see a lot of really strong evidence that it is the correct thing to do.

**Reply by hillin** [reply]
I haven't read iter-tools' implementation so it's for not me to say whether it's simple; and again I agree it's only something worth trying if there is *a lot of really strong evidences*, which are *profiling shows the step object allocation/collection is a hot code path*, *it's not clear how and when (and whether) would the JS engines optimize it* and *there is no other better optimization choice*.

**Reply by conartist6** [reply]
OK, so I'm telling you that I have spent time thinking about this and investigating and while I could not say that all my related investigations are complete, I'm thinking that those conditions are not met yet. Step allocation and GC are both hot, but known to be well optimized. Engine implementers have strong incentive to optimize these patterns as they are now fundamental to the language.

**Comment by conartist6** [comment]
One idea is to simply do nothing. That is to say, all the patterns iter-tools uses should be possible to optimize further at the engine level, and if people start using iterators and generators more then it will only be a matter of time until engines prioritize the optimizations relevant to them. The advantage of this approach is that it is pretty sure to work given enough time, and anything else I do to optimize is likely to eventually be beaten by native optimizations.

Some native optimization work is out there already, for example v8 already turns `[].map().filter().map()` into an iterative algorithm internally to avoid intermediate allocations.

**Comment by conartist6** [comment]
One idea is to transpile generators into classes. A problem we've seen is that it's pretty expensive to pack up all the execution state of a function and restore it. If we transpiled to classes we could have a regular old function as a `next()` method. Instead of having to pack up values from each in-scope symbol and stash them away (`t*n` times) the state of each iterator would simply be stored in class properties.

This would combine well with eliminating step objects since that approach would essentially require this one as a prerequisite.
